File: Physics/PhysicsBasedAnimation/implicitFem.py
import numpy as np
import scipy.linalg
# 已完成，数据完全一致，稍微有点精度问题

dt = 0.00033333
lam = 1e4
mu = 1e7
damping = 0.01
density = 1e4

'''    读取文件  开始    '''

f = open("code/inputs/input.mesh","r")   #设置文件对象
line = f.readline()
num_elem = 0
num_particle = 1
num_tri = 0
while line:   
   line = f.readline()
   st = line.split(' ')
   if st[0] == 'p':
       num_particle += 1
   elif st[0] == 'e':
       num_elem += 1
   elif st[0] == 't':
       num_tri += 1
       
f = open("code/inputs/input.mesh","r")   #设置文件对象
pos = np.zeros((num_particle,3))
force = np.zeros((num_particle,3))
vel = np.zeros((num_particle,3))
mass = np.zeros((num_particle))


normal = np.zeros((num_elem,4,3))
elementQ = np.zeros((num_elem,3,3))
stiffness = np.zeros((num_elem,16,9))
elements = np.zeros((num_elem,4),dtype = int)
basis = np.zeros((num_elem,3,3))
tri = np.zeros((num_tri,3),dtype = int)
pcnt = 0
ecnt = 0
tcnt = 0
line = f.readline()
st = line.split(' ')
pos[pcnt,0] = st[1]
pos[pcnt,1] = st[2]
pos[pcnt,2] = st[3]
pcnt += 1
while line:   
   line = f.readline()
   st = line.split(' ')
   if st[0] == 'p':
       pos[pcnt,0] = st[1]
       pos[pcnt,1] = st[2]
       pos[pcnt,2] = st[3]
       pcnt += 1
   elif st[0] == 'e':
       elements[ecnt,0] = st[1]
       elements[ecnt,1] = st[2]
       elements[ecnt,2] = st[3]
       elements[ecnt,3] = st[4]
       ecnt += 1
   elif st[0] == 't':
       tri[tcnt,0] = st[1]
       tri[tcnt,1] = st[2]
       tri[tcnt,2] = st[3]
       tcnt += 1
       
f.close() #关闭文件
'''    读取文件  结束    '''

def cross(u,v):
    t0 = u[1]*v[2] - u[2]*v[1]
    t1 = u[2]*v[0] - u[0]*v[2]
    t2 = u[0]*v[1] - u[1]*v[0]
    return np.array([t0,t1,t2])

def scaledot(mat,vec):
    res = np.zeros((3))
    for i in range(3):
        for j in range(3):
            res[i] += mat[i,j] * vec[j]
    return res

def vecdot(vec0,vec1):
    res = 0
    for i in range(len(vec0)):
        res += vec0[i] * vec1[i]
    return res

def init():
    for ie in range(elements.shape[0]):
        x0 = pos[elements[ie,0],:]
        x1 = pos[elements[ie,1],:]
        x2 = pos[elements[ie,2],:]
        x3 = pos[elements[ie,3],:]
    
        X = np.zeros((3,3))
        X[0,:] = x1 - x0
        X[1,:] = x2 - x0
        X[2,:] = x3 - x0
        
        normal[ie,0,:] = cross(x2 - x1,x3 - x1)
        normal[ie,1,:] = cross(x3 - x0,x2 - x0)
        normal[ie,2,:] = cross(x1 - x0,x3 - x0)
        normal[ie,3,:] = cross(x2 - x0,x1 - x0)
        
        det = np.linalg.det(X)
        m = density * det / 24
        basis[ie,:,:] = np.linalg.inv(X)
        mass[elements[ie,0]] += m
        mass[elements[ie,1]] += m
        mass[elements[ie,2]] += m
        mass[elements[ie,3]] += m
        
        for i in range(4):
            for j in range(4):
                ni = normal[ie,i,:]
                nj = normal[ie,j,:]
                nnt = np.zeros((3,3))
                for i0 in range(3):
                    for j0 in range(3):
                        nnt[i0,j0] = ni[i0] * nj[j0]
                mat = np.zeros((3,3))
                idx = int(i * 4 + j)
                mat = lam * nnt + mu * np.transpose(nnt)
                diag = mu * vecdot(ni,nj)
                mat[0,0] += diag
                mat[1,1] += diag
                mat[2,2] += diag
                mat /= (6 * det)
                
                for j0 in range(3):
                    for i0 in range(3):
                        idx0 = int(j0 * 3 + i0)
                        stiffness[ie,idx,idx0] = mat[i0,j0]
             
def dotMat(mat0,mat1):
    n = mat0.shape[0]
    m = mat0.shape[1]
    res = 0
    for i in range(n):
        for j in range(m):
           res += mat0[i,j] * mat1[i,j] 
    return res

def solve(b,x):
    # 又是共轭梯度，我都快背下来了
    tol = 1e-4
    ite = 0
    ite_max = 100
    r = np.zeros((num_particle,3))
    d = np.zeros((num_particle,3))
    q = np.zeros((num_particle,3))
    
    mvmul(x,q)
    
    r = b - q
    d = r.copy()
    
    sigma = dotMat(r,r)
    tol *= sigma
    
    while(ite < ite_max):
        ite += 1
        mvmul(d, q)
        alpha = sigma / dotMat(d,q)
        x += alpha * d
        r -= alpha * q
        sigma_old = sigma
        sigma = dotMat(r,r)
        if sigma < tol or alpha < 1e-20:
            break
        beta = sigma / sigma_old
        q = r.copy()
        q += beta * d
        d = q.copy()
        
        
    
def mvmul(x,y):
    global mass
    for p in range(num_particle):
        y[p,:] = mass[p] * x[p,:]
    scale = dt * dt + dt * damping
    for ie in range(num_elem):
        for i in range(4):
            for j in range(4):
                
                mat = np.zeros((3,3))
                idx = int(4 * i + j)
                for j0 in range(3):
                    for i0 in range(3):
                        idx0 = j0 * 3 + i0
                        mat[i0,j0] = stiffness[ie,idx,idx0]
                        
                term = np.dot(np.dot(elementQ[ie,:,:],mat),np.transpose(elementQ[ie,:,:]))
                y[elements[ie,i],:] += scale * scaledot(term[:,:],x[elements[ie,j],:])
init()
time = 0
timeFinal = 8


while(time < timeFinal):
    
    for p in range(num_particle):
        force[p,:] = 0
        force[p,2] = -9.8 * mass[p]
    
    for ie in range(elements.shape[0]):
        x0 = pos[elements[ie,0],:]
        x1 = pos[elements[ie,1],:]
        x2 = pos[elements[ie,2],:]
        x3 = pos[elements[ie,3],:]
    
        v0 = vel[elements[ie,0],:]
        v1 = vel[elements[ie,1],:]
        v2 = vel[elements[ie,2],:]
        v3 = vel[elements[ie,3],:]
        X = np.zeros((3,3))
        X[0,:] = x1 - x0
        X[1,:] = x2 - x0
        X[2,:] = x3 - x0
        
        # 注意Eigen库和numpy算乘法的方式不一样，不过怎么会不一样呢？
        F = np.dot(basis[ie],X)
        
        # scipy 的 svd 解法的精度问题
        for i in range(3):
            for j in range(3):
                if F[i,j] < 1e-10:
                    F[i,j] = 0
        Q = np.zeros((3,3)) # 旋转矩阵
        # 注意scipy 算出来的是V的转置，所以是V_transpose
        U,sigma,Vt = scipy.linalg.svd(F)
        
        Q = np.dot(Vt,U)
        Ftilde = np.dot(F,np.transpose(Q))
        
        iden = np.zeros((3,3))
        iden[0,0] = iden[1,1] = iden[2,2] = 1
        strain = (Ftilde + np.transpose(Ftilde)) / 2 - iden
        tr = strain[0,0] + strain[1,1] + strain[2,2]
        stress = lam * tr * iden + 2 * mu * strain

        qs = np.dot(stress,Q)
        
        elementQ[ie,:,:] = Q[:,:]
        
        force[elements[ie,0],:] += scaledot(qs,normal[ie,0,:]) / 6
        force[elements[ie,1],:] += scaledot(qs,normal[ie,1,:]) / 6
        force[elements[ie,2],:] += scaledot(qs,normal[ie,2,:]) / 6
        force[elements[ie,3],:] += scaledot(qs,normal[ie,3,:]) / 6
        
    x = np.zeros((num_particle,3))
    b = np.zeros((num_particle,3))
    
    for p in range(num_particle):
        b[p,:] = mass[p] * vel[p,:] + dt * force[p,:]
        x[p,:] = vel[p,:]
        
    solve(b,x)
        
    for p in range(num_particle):
        vel[p,:] = x[p,:]
        pos[p,:] += dt * vel[p,:]
        
        if(pos[p,2] < 0):
            pos[p,2] = 0
            vel[p,2] = 0
        

File: Physics/PhysicsBasedAnimation/code/implicitfem.cpp
#include "implicitfem.h"

static const double FEPS = 1e-6;

void polar(const Eigen::Matrix3d &F, Eigen::Matrix3d &Q) {
    Eigen::JacobiSVD<Eigen::Matrix3d> svd(F, Eigen::ComputeFullU | Eigen::ComputeFullV);
    Eigen::Matrix3d U = svd.matrixU();
    Eigen::Matrix3d V = svd.matrixV();
    Q = U * V.transpose();
}

void initialize(Object &obj, const SimulationParameters &params) {
  for (std::vector<Particle>::iterator p = obj.particles.begin(); p != obj.particles.end(); p++) {
    p->mass = 0.0;
    p->rest = p->pos;
  }

  for (std::vector<Element>::iterator e = obj.elements.begin(); e != obj.elements.end(); e++) {
    Eigen::Vector3d &x0 = obj.particles[(*e)[0]].pos;
    Eigen::Vector3d &x1 = obj.particles[(*e)[1]].pos;
    Eigen::Vector3d &x2 = obj.particles[(*e)[2]].pos;
    Eigen::Vector3d &x3 = obj.particles[(*e)[3]].pos;

    e->normals[0] = (x2-x1).cross(x3-x1);
    e->normals[1] = (x3-x0).cross(x2-x0);
    e->normals[2] = (x1-x0).cross(x3-x0);
    e->normals[3] = (x2-x0).cross(x1-x0);

    e->basis << x1-x0, x2-x0, x3-x0;
    double det = e->basis.determinant();
    double mass = params.density * det / 24.0;
    e->basis = e->basis.inverse().eval();

    obj.particles[(*e)[0]].mass += mass;
    obj.particles[(*e)[1]].mass += mass;
    obj.particles[(*e)[2]].mass += mass;
    obj.particles[(*e)[3]].mass += mass;

    for (unsigned int i=0; i<4; i++) {
      for (unsigned int j=0; j<4; j++) {
	Eigen::Vector3d &ni = e->normals[i];
	Eigen::Vector3d &nj = e->normals[j];
	Eigen::Matrix3d nnt;
	nnt << ni[0]*nj[0], ni[0]*nj[1], ni[0]*nj[2],
	  ni[1]*nj[0], ni[1]*nj[1], ni[1]*nj[2],
	  ni[2]*nj[0], ni[2]*nj[1], ni[2]*nj[2];
	Eigen::Matrix3d &m = e->stiffness[4*i+j];
	m = params.lambda*nnt + params.mu*nnt.transpose();
	double diag = params.mu * ni.dot(nj);
	m(0,0) += diag;
	m(1,1) += diag;
	m(2,2) += diag;
	m /= 6.0 * det;
      }
    }
  }
}

void mvmul(const SimulationParameters &params, const Object &obj, const Eigen::Vector3d *x, Eigen::Vector3d *y) {
  double scale = params.dt*params.dt + params.dt*params.damp;

  for (unsigned int i=0; i<obj.particles.size(); i++) {
    y[i] = obj.particles[i].mass * x[i];
  }
  
  for (std::vector<Element>::const_iterator e=obj.elements.begin(); e != obj.elements.end(); e++) {
    for (unsigned int i=0; i<4; i++) {
      for (unsigned int j=0; j<4; j++) {
	y[(*e)[i]] += scale * e->Q * e->stiffness[4*i+j] * e->Q.transpose() * x[(*e)[j]];
      }
    }
  }
}

void daxpy(int n, double alpha, Eigen::Vector3d *x, Eigen::Vector3d *y) {
  for (unsigned int i=0; i<n; i++) y[i] += alpha*x[i];
}

double ddot(int n, Eigen::Vector3d *x, Eigen::Vector3d *y) {
  double val = 0;
  for (unsigned int i=0; i<n; i++) val += x[i].dot(y[i]);
  return val;
}
void dcopy(int n, Eigen::Vector3d *x, Eigen::Vector3d *y) {
  for (unsigned int i=0; i<n; i++) y[i] = x[i];
}


void solve(const SimulationParameters &params, const Object &obj, const Eigen::Vector3d *b, Eigen::Vector3d *x) {
  double tol = 1e-4;
  int iterations = 0;
  int n = obj.particles.size();
  Eigen::Vector3d *r = new Eigen::Vector3d[n];
  Eigen::Vector3d *d = new Eigen::Vector3d[n];
  Eigen::Vector3d *q = new Eigen::Vector3d[n];

  mvmul(params, obj, x, q);

  for (unsigned int i=0; i<n; i++) {
    r[i] = b[i] - q[i];
    d[i] = r[i];
  }

  double sigma = ddot(n, r, r);
  tol *= sigma;

  while (iterations++ < 100) {
    mvmul (params, obj, d, q);
    double alpha = sigma / ddot(n, d, q);
    daxpy(n, alpha, d, x);
    daxpy(n, -alpha, q, r);
    double sigma_old = sigma;
    sigma = ddot(n, r, r);
    if (sigma < tol || alpha < 1e-20) break;
    double beta = sigma / sigma_old;
    dcopy(n, r, q);
    daxpy(n, beta, d, q);
    dcopy(n, q, d);
  }
  
  delete [] r;
  delete [] d;
  delete [] q;
}

int main(int argc, char *argv[]) {
  char fname[80];
  SimulationParameters params;
  std::vector<Object> objects;
  readInputFile(argv[1], params, objects);

  double time = 0;
  int frame = 0;
  double frameTime = -1.0;

  for (std::vector<Object>::iterator obj = objects.begin(); obj != objects.end(); obj++) {
    initialize(*obj, params);
  }
  
  while (time < params.total_time) {
    // write frame
    if (frameTime < 0) {
      for (unsigned int o=0; o<objects.size(); o++) {
	sprintf(fname, params.output_fname.c_str(), o, frame);
	writeObj(fname, objects[o].particles, objects[o].triangles);
	frameTime = 1.0/30.0-0.0001;
	frame++;
      }
    }

    // takes a timestep
    for (std::vector<Object>::iterator obj = objects.begin(); obj != objects.end(); obj++) {
      // initialize force accumulator, apply gravity
      for (std::vector<Particle>::iterator p = obj->particles.begin(); p != obj->particles.end(); p++) {
	p->frc = Eigen::Vector3d::Zero();
	p->frc[2] = -9.8*p->mass;
      }

      // comupte and accumulate spring forces
      for (std::vector<Element>::iterator e = obj->elements.begin(); e != obj->elements.end(); e++) {
	Eigen::Matrix3d X, F, Q, Ftilde, strain, stress;
	Eigen::Vector3d &x0 = obj->particles[(*e)[0]].pos;
	Eigen::Vector3d &x1 = obj->particles[(*e)[1]].pos;
	Eigen::Vector3d &x2 = obj->particles[(*e)[2]].pos;
	Eigen::Vector3d &x3 = obj->particles[(*e)[3]].pos;

	X << x1-x0, x2-x0, x3-x0;
	F = X * e->basis;
	polar(F, Q);
	Ftilde = Q.transpose() * F;
	strain = 0.5 * (Ftilde + Ftilde.transpose()) - Eigen::Matrix3d::Identity();
	stress = params.lambda * strain.trace() * Eigen::Matrix3d::Identity() + 2 * params.mu * strain;
	
	obj->particles[(*e)[0]].frc += Q * stress * e->normals[0] / 6.0;
	obj->particles[(*e)[1]].frc += Q * stress * e->normals[1] / 6.0;
	obj->particles[(*e)[2]].frc += Q * stress * e->normals[2] / 6.0;
	obj->particles[(*e)[3]].frc += Q * stress * e->normals[3] / 6.0;

	e->Q = Q;
      }

      // time integration
      Eigen::Vector3d *x = new Eigen::Vector3d[obj->particles.size()];
      Eigen::Vector3d *b = new Eigen::Vector3d[obj->particles.size()];
      int i=0;
      for (std::vector<Particle>::iterator p = obj->particles.begin(); p != obj->particles.end(); p++, i++) {
	b[i] = p->mass * p->vel + params.dt * p->frc;
	x[i] = p->vel;
      }

      solve(params, *obj, b, x);
      
      i=0;
      for (std::vector<Particle>::iterator p = obj->particles.begin(); p != obj->particles.end(); p++, i++) {
	p->vel = x[i];
	p->pos += params.dt*(p->vel);

	// ground collision
	if (p->pos[2] < 0.0) {
	  p->pos[2] = 0.0;
	  p->vel[2] = 0.0;
	}
      }
      delete [] x;
      delete [] b;
    }
    
    time += params.dt;
    frameTime -= params.dt;
  }
}














/////////////////////////////////////////
// I/O
/////////////////////////////////////////

#include "json/json.h"
#include <fstream>

bool readInputFile(const char *fname, 
    SimulationParameters &params,
    std::vector<Object> &objects) {
  
  std::ifstream in(fname, std::ios::in);
  
  Json::Value root;
  Json::Reader jReader;
  
  if(!jReader.parse(in, root)){
    std::cout << "couldn't read input file: " << fname << '\n'
	      << jReader.getFormattedErrorMessages() << std::endl;
    exit(1);
  }
  
  params.dt = root.get("dt", 1.0/300.0).asDouble();
  params.total_time = root.get("total_time", 1.0).asDouble();
  params.density = root.get("density", 1e4).asDouble();
  params.lambda = root.get("lambda", 1e4).asDouble();
  params.mu = root.get("mu", 1e4).asDouble();
  params.damp = root.get("damping", 1e-2).asDouble();
  params.output_fname = root.get("output_fname", std::string("output-%02d.%04d.obj")).asString();
  
  Json::Value objectsIn = root["objects"];
  objects.resize(objectsIn.size());
  for (unsigned int i=0; i<objectsIn.size(); i++) {
    readObject((objectsIn[i]["filename"]).asString().c_str(), objects[i]);
  }
  return true;
}

bool readObject(const char *fname, Object &object) {
  char ch;
  Particle p;
  Element e;
  Tri t;
  

  std::ifstream in(fname, std::ios::in);
  while (in>>ch) {
    if (ch == 'p') {
      in>>p.pos[0]>>p.pos[1]>>p.pos[2]>>p.vel[0]>>p.vel[1]>>p.vel[2];
      object.particles.push_back(p);
      continue;
    }
    if (ch == 'e') {
      in>>e[0]>>e[1]>>e[2]>>e[3];
      object.elements.push_back(e);
      continue;
    }
    if (ch == 't') {
      in>>t[0]>>t[1]>>t[2];
      object.triangles.push_back(t);
      continue;
    }
  }
  in.close();
  std::cout<<"inputfile " << fname <<" read"<<std::endl;
  return true;
}

void writeObj(char *fname, const std::vector<Particle> &meshPts, const std::vector<Tri> &triangles) {
  std::cout<<"writing "<<fname<<std::endl;
  std::ofstream out;
  std::vector<Particle>::const_iterator p;
  std::vector<Tri>::const_iterator t;
  
  out.open(fname);
  
  for (p=meshPts.begin(); p!=meshPts.end(); p++) 
    out<<"v "<<p->pos[0]<<" "<<p->pos[1]<<" "<<p->pos[2]<<std::endl;
  
  for (t=triangles.begin(); t!=triangles.end(); t++) 
    out<<"f "<<(*t)[0]+1<<" "<<(*t)[1]+1<<" "<<(*t)[2]+1<<std::endl;
  
  out.close();
}


